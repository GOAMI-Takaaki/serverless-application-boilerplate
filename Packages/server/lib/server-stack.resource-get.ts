import {DynamoDB} from 'aws-sdk';
import {APIGatewayEventRequestContext, APIGatewayProxyEvent, APIGatewayProxyResult} from 'aws-lambda';

const db = new DynamoDB.DocumentClient();

export async function handler (
  event: APIGatewayProxyEvent,
  context: APIGatewayEventRequestContext
): Promise<APIGatewayProxyResult> {
  const resourceType = event.pathParameters?.resourceType;
  const resourceId = event.pathParameters?.resourceId;
  if (!resourceType||!resourceId) {
    return {
      statusCode: 400,
      body: `Error: You are missing the path parameter resource type or id`,
    };
  }
  
  const params = {
    TableName: resourceType,
    Key: {
        id: resourceId,
      },
  };
  
  try {
    const response = await db.get(params).promise();
      return { statusCode: 200, body: JSON.stringify(response.Item) };
    } catch (dbError) {
      return { statusCode: 500, body: JSON.stringify(dbError) };
  }
}