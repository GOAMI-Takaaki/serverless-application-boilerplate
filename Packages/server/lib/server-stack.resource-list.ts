import {DynamoDB} from 'aws-sdk';
import {APIGatewayEventRequestContext, APIGatewayProxyEvent, APIGatewayProxyResult} from 'aws-lambda';

const db = new DynamoDB.DocumentClient();

export async function handler (
  event: APIGatewayProxyEvent,
  context: APIGatewayEventRequestContext
): Promise<APIGatewayProxyResult> {
  const resourceType = event.pathParameters?.resourceType;
    if (!resourceType) {
      return {
        statusCode: 400,
        body: `Error: You are missing the path parameter resource type`,
      };
    }

  // TOODO
  const params = {
    TableName: resourceType,
    ExpressionAttributeValues: {
      ':s': {N: '2'},
      ':e' : {N: '09'},
      ':topic' : {S: 'PHRASE'}
    },
    ProjectionExpression: 'Episode, Title, Subtitle',
    KeyConditionExpression: 'Season = :s and Episode > :e',
    FilterExpression: 'contains (Subtitle, :topic)',
  };
  
  // const params2 = {
  //   TableName: resourceType,
  //   ExpressionAttributeValues: {
  //     ":topic": {S: "SubTitle2"},
  //     ":s": {N: 1},
  //     ":e": {N: 2},
  //   },
  //   ProjectionExpression: "Season, Episode, Title, Subtitle",
  //   FilterExpression: "Subtitle = :topic AND Season = :s AND Episode = :e",
  // };

  try {
    const response = await db.query(params).promise();
      return { statusCode: 200, body: JSON.stringify(response.Items) };
    } catch (dbError) {
      return { statusCode: 500, body: JSON.stringify(dbError) };
  }
}