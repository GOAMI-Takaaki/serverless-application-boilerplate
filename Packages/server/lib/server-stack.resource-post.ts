import {DynamoDB} from 'aws-sdk';
import {APIGatewayEventRequestContext, APIGatewayProxyEvent, APIGatewayProxyResult} from 'aws-lambda';
import { randomUUID } from 'crypto';

const db = new DynamoDB.DocumentClient();

export async function handler (
  event: APIGatewayProxyEvent,
  context: APIGatewayEventRequestContext
): Promise<APIGatewayProxyResult> {
    const resourceType = event.pathParameters?.resourceType;
      if (!resourceType) {
        return {
          statusCode: 400,
          body: `Error: You are missing the path parameter resource type`,
        };
      }
  
    const params = {
      TableName: resourceType,
      Item: {
          id: randomUUID(),
        },
    };
  
    try {
      await db.put(params).promise();
        return { statusCode: 200, body: JSON.stringify(params.Item) };
      } catch (dbError) {
        return { statusCode: 500, body: JSON.stringify(dbError) };
    }
}