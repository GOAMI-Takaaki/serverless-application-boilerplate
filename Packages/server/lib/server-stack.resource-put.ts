import {DynamoDB} from 'aws-sdk';
import {APIGatewayEventRequestContext, APIGatewayProxyEvent, APIGatewayProxyResult} from 'aws-lambda';

const db = new DynamoDB.DocumentClient();

export async function handler (
  event: APIGatewayProxyEvent,
  context: APIGatewayEventRequestContext
): Promise<APIGatewayProxyResult> {
  const resourceType = event.pathParameters?.resourceType;
  const resourceId = event.pathParameters?.resourceId;
  if (!resourceType||!resourceId) {
    return {
      statusCode: 400,
      body: `Error: You are missing the path parameter resource type or id`,
    };
  }
  
  const params = {
    TableName: resourceType,
    Item: {
        id: resourceId,
      },
  };
  
  try {
    await db.put(params).promise();
      return { statusCode: 200, body: JSON.stringify(params.Item) };
    } catch (dbError) {
      return { statusCode: 500, body: JSON.stringify(dbError) };
  }
}