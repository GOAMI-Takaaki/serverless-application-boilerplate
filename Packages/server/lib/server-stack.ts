import * as cdk from '@aws-cdk/core';
import { Table, AttributeType } from '@aws-cdk/aws-dynamodb';
import { RestApi, ApiKeySourceType, Cors,LambdaIntegration, Resource } from '@aws-cdk/aws-apigateway';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import { IFunction, Runtime } from '@aws-cdk/aws-lambda';


export class ServerStack extends cdk.Stack {

  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.createApis();
    this.createTable("items");
  }

  createApis(): void {
    const api = new RestApi(this, "reource-api", {
      restApiName: "reource-api",
      apiKeySourceType: ApiKeySourceType.HEADER,
      defaultCorsPreflightOptions: {
        allowOrigins: Cors.ALL_ORIGINS,
        allowMethods: Cors.ALL_METHODS,
        allowHeaders: Cors.DEFAULT_HEADERS,
        statusCode: 200,
      }
    });
  
    const resourceType = api.root.addResource("{resourceType}");
    this.addResourceMethod(resourceType, "GET", 'resource-list');
    this.addResourceMethod(resourceType, "POST", 'resource-post');

    const resource = resourceType.addResource("{resourceId}");
    this.addResourceMethod(resource, "GET", 'resource-get');
    this.addResourceMethod(resource, "PUT", 'resource-put');
    this.addResourceMethod(resource, "DELETE", 'resource-delete');
  }

  addResourceMethod(resource: Resource, method: string, functionName: string): void {
    const nodejsFunction = new NodejsFunction(this, functionName, {
      functionName: functionName,
      runtime: Runtime.NODEJS_14_X,
      timeout: cdk.Duration.seconds(30),
    });
    resource.addMethod(method, new LambdaIntegration(nodejsFunction), {
      apiKeyRequired: true,
    });
  }

  createTable(resourceType: string): void {
    new Table(this, `table-${resourceType}`, {
      partitionKey: {
        name: "id",
        type: AttributeType.STRING,
      },
      tableName: resourceType,
      removalPolicy: cdk.RemovalPolicy.DESTROY, // TODO swtich by stage
    });
  }
}
