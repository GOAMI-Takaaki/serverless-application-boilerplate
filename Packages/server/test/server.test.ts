import { expect as expectCDK, matchTemplate, MatchStyle, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as Server from '../lib/server-stack';

test('Confirm snapshot', () => {
  const app = new cdk.App();
  // WHEN
  const stack = new Server.ServerStack(app, 'TestStack');
  // THEN
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
