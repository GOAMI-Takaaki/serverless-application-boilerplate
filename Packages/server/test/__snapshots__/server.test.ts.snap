// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Confirm snapshot 1`] = `Object {}`;
